packages/langium/src/grammar/langium-grammar-validator.ts

```diff
@@ -112,6 +117,69 @@ export class LangiumGrammarValidator {
+        const importedRules = new Set<string>();
+        const resolvedGrammars = resolveTransitiveImports(this.documents, grammar);
+        for (const resolvedGrammar of resolvedGrammars) {
+            for (const rule of resolvedGrammar.rules) {
+                importedRules.add(rule.name.toLowerCase());
+            }
+        }
+        for (const name of ruleMap.keys()) {
+            if (importedRules.has(name)) {
+                const rules = ruleMap.get(name);
+                rules.forEach(e => {
+                    accept('error', `A rule with the name '${e.name}' already exists in an imported grammar.`, { node: e, property: 'name' });
+                });
+            }
+        }
+    }
+
+    /**
+     * Compared to the validation above, this validation only checks whether two imported grammars export the same grammar rule.
+     */
+    checkUniqueImportedRules(grammar: ast.Grammar, accept: ValidationAcceptor): void {
+        const imports = new Map<ast.GrammarImport, ast.Grammar[]>();
+        for (const imp of grammar.imports) {
+            const importedGrammars = resolveTransitiveImports(this.documents, imp);
+            imports.set(imp, importedGrammars);
+        }
+        const allDuplicates = new MultiMap<ast.GrammarImport, string>();
+        for (const outerImport of grammar.imports) {
+            const outerGrammars = imports.get(outerImport)!;
+            for (const innerImport of grammar.imports) {
+                if (outerImport === innerImport) {
+                    continue;
+                }
+                const innerGrammars = imports.get(innerImport)!;
+                const duplicates = this.getDuplicateExportedRules(outerGrammars, innerGrammars);
+                for (const duplicate of duplicates) {
+                    allDuplicates.add(outerImport, duplicate);
+                }
+            }
+        }
+        for (const imp of grammar.imports) {
+            const duplicates = allDuplicates.get(imp);
+            if (duplicates.length > 0) {
+                accept('error', 'Some rules exported by this grammar are also included in other imports: ' + stream(duplicates).distinct().join(', '), { node: imp, property: 'path' });
+            }
+        }
+    }
+
+    private getDuplicateExportedRules(outer: ast.Grammar[], inner: ast.Grammar[]): Set<string> {
+        const exclusiveOuter = outer.filter(g => !inner.includes(g));
+        const outerRules = exclusiveOuter.flatMap(e => e.rules);
+        const innerRules = inner.flatMap(e => e.rules);
+        const duplicates = new Set<string>();
+        for (const outerRule of outerRules) {
+            const outerName = outerRule.name.toLowerCase();
+            for (const innerRule of innerRules) {
+                const innerName = innerRule.name.toLowerCase();
+                if (outerName === innerName) {
+                    duplicates.add(innerRule.name);
+                }
+            }
+        }
+        return duplicates;
```

Comment:
This check can potentially be computation-heavy for large projects. It's ok for now, but let's keep the performance aspect in mind.